Account for multiple modules when looking up the DeclContext of a type

augusto2112 opened the pull request:
When looking up the decl context of a type, ASTDemangler has to take into account that there are multiple different modules where that type could've come from. This is due to two facts:

- Thanks to the `-module-abi-name` flag, multiple modules can share the same ABI name (which is the module name that is usually used when mangling a type).
- In some situations mangling can use the module's real name, for example, when mangling for the debugger or USRs coupled with @_originallyDefinedIn.

rdar://134095412

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci smoke test

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -2713,12 +2713,16 @@ ModuleDecl *ASTContext::getModuleByIdentifier(Identifier ModuleID) {
+ASTContext::getModulesByRealOrABIName(StringRef ModuleName) {
+  llvm::SmallVector<ModuleDecl *, 1> modules;
+
+  for (auto &[_, module] : getLoadedModules())
```
the number of loaded modules can be large. Instead why don't you pre-populate a mapping from module name to a list of modules with that name

adrian-prantl replied:
Is "RealName" used elsewhere? If it's a new term, may I suggest "ASTName"?

augusto2112 replied:
That was my initial idea, however the module's ABI name is mutable (there's a setter for it), so we could easily end up in a situation where keeping that mapping and the module's ABI name are out of sync...

augusto2112 replied:
> Is "RealName" used elsewhere? If it's a new term, may I suggest "ASTName"?

`RealName` is an actual term already.

adrian-prantl replied:
> however the module's ABI name is mutable

I thought this gets set via `-module-abi-name` and is global? Is there a legitimate use-case for setting the ABI name in the middle of a compilation?

augusto2112 replied:
Yes it's set via `-module-abi-name` and global, what I mean is, on `ModuleDecl` there's a `setABIName` method which is called at some point by the module loader/frontend. I can implement a working solution with the mapping now, but my worry is that in the future it could be quite easy to add a call to `setABIName` somewhere else and make it so the mapping is out of sync.

augusto2112 replied:
Actually it looks like the `ModuleDecl` has access to the AST it's in, so I should be able to do this in a way that avoids problems in the future.

slavapestov commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -264,25 +264,26 @@ Type ASTBuilder::resolveOpaqueType(NodePointer opaqueDescriptor,
+      auto opaqueDecl = module->lookupOpaqueResultType(mangledName);
+      if (!opaqueDecl)
+        continue;
+      SmallVector<Type, 8> allArgs;
```
It's confusing to stick the rest of the body here inside the loop, even though it will only run once.

slavapestov commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -1321,39 +1340,43 @@ ASTBuilder::findDeclContext(NodePointer node) {
-        if (!genericSig || genericSigMatchesNominal)
-          return ext;
+    for (auto *moduleDecl : moduleDecls) {
+      for (auto *ext : nominalDecl->getExtensions()) {
```
Again let's avoid a double nested loop here by changing the `if (ext->getParentModule() != moduleDecl)` to an `if (std::find(...))` or similar, and keeping the rest of the identation as-is

augusto2112 committed:
Account for multiple modules when looking up the DeclContext of a type

When looking up the decl context of a type, ASTDemangler has to take
into account that there are multiple different modules where that type
could've come from. This is due to two facts:

- Thanks to the `-module-abi-name` flag, multiple modules can share
the same ABI name (which is the module name that is usually used when
mangling a type).
- In some situations mangling can use the module's real name, for
example, when mangling for the debugger or USRs coupled with @_originallyDefinedIn.

rdar://134095412

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci smoke test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@slavapestov ping :)

adrian-prantl commented on include/swift/AST/ASTContext.h:
```diff
@@ -1200,10 +1200,16 @@ class ASTContext final {
   ///
-  /// \returns The module if found, nullptr otherwise.
-  ModuleDecl *getLoadedModuleByABIName(StringRef ModuleName);
+  /// Modules that are being looked up by ABI name must be a dependency
```
```suggestion
  /// Modules that are being looked up by ABI name are only found if they are a dependency
```

adrian-prantl commented on include/swift/AST/ASTContext.h:
```diff
@@ -1200,10 +1200,16 @@ class ASTContext final {
+  /// of a module that has that same name as its real name, as there is no
+  /// efficient way to lazily load a module by ABI name.
+  llvm::SmallVector<ModuleDecl *, 1> *
+  getModulesByRealOrABIName(StringRef ModuleName);
```
The API returning a pointer to a SmallVector is odd. From a quick look at the implementation — couldn't this be an llvm::ArrayRef<ModuleDecl *> ?

adrian-prantl commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -1237,20 +1244,24 @@ ASTBuilder::findDeclContext(NodePointer node) {
-      if (!decl) return nullptr;
+      for (auto *module : *potentialModules)
+        if (auto *decl = module->lookupLocalType(mangledName))
+          return dyn_cast<DeclContext>(decl);
```
Could this be a findModuleContaining(mangledName) helper function that is reused?

augusto2112 replied:
I could put this in a helper but besides iterating over the modules there's nothing in common I could reuse between callers of `findPotentialModules`.

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci smoke test
